File: leetcode/0001-1000/0801-0900/0827-making-a-large-island/MakingALargeIsland.ts
class Island {
  constructor(public id: number, public area: number) { }
}

const directions = [[-1, 0], [0, 1], [1, 0], [0, -1]];

const cell2island = (grid: number[][]): Map<number, Island> => {
  const n = grid.length;
  const islands = new Map();
  const visited: boolean[][] = new Array(n).fill(false).map(
    (_) => new Array(n).fill(false)
  );

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      if (visited[i][j]) {
        continue;
      }

      if (grid[i][j] === 0) {
        visited[i][j] = true;
        continue;
      }

      // Found an 1 cell
      const queue = [{ x: i, y: j, island: new Island(i * n + j, 0) }];

      while (queue.length > 0) {
        //
        const { x, y, island } = queue.shift()!;

        if (visited[x][y]) {
          continue;
        }

        visited[x][y] = true;

        if (grid[x][y] === 0) {
          continue;
        }

        const position = x * n + y;
        island.area++;

        islands.set(position, island);

        // Prepare for the next queue
        for (const [dx, dy] of directions) {
          const newX = x + dx;
          const newY = y + dy;

          if (newX < 0 || newX >= n || newY < 0 || newY >= n) {
            continue;
          }

          if (visited[newX][newY]) {
            continue;
          }

          //
          queue.push({ x: newX, y: newY, island });
        }
      }
    }
  }

  return islands;
}

function largestIsland(grid: number[][]): number {

  const islands = cell2island(grid);

  let max = Math.max(...[...islands.values()].map(c => c.area));

  const n = grid.length;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] === 0) {
        //
        const connected = directions.map(([x, y]) => [x + i, y + j])
          .filter(([x, y]) => x >= 0 && x < n && y >= 0 && y < n && grid[x][y] === 1)
          // .map(([x, y]) => x * n + y)
          .map(([x, y]) => islands.get(x * n + y)!);

        max = Math.max(max, 1 + [...new Set(connected)].reduce(
          (accumulated, value) => accumulated + value.area, 0
        ))
      }
    }
  }

  return max;
};


console.log(largestIsland([
  [1, 1, 0, 1],
  [1, 1, 0, 1],
  [1, 1, 0, 1],
  [1, 1, 0, 1],
]));

File: leetcode/0001-1000/0801-0900/0827-making-a-large-island/MakingALargeIsland.java
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

public class MakingALargeIsland {
  public static void main(String[] args) {
    int[][] grid = {
        { 1, 1, 0, 1 },
        { 1, 1, 0, 1 },
        { 1, 1, 0, 1 },
        { 1, 1, 0, 1 }
    };

    System.out.println(new Solution().largestIsland(grid));
  }
}

class Solution {
  public int largestIsland(int[][] grid) {
    int n = grid.length;
    boolean[][] visited = new boolean[n][n];
    int[][] directions = {
        { -1, 0 },
        { 0, 1 },
        { 1, 0 },
        { 0, -1 }
    };

    Map<Integer, Integer> islands = new HashMap<>();

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        if (visited[i][j]) {
          continue;
        }

        if (grid[i][j] == 0) {
          visited[i][j] = true;
          continue;
        }

        int id = i * n + j + 1;
        int area = 0;

        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[] { i, j });

        while (!queue.isEmpty()) {
          int[] poll = queue.poll();

          int x = poll[0];
          int y = poll[1];

          if (visited[x][y]) {
            continue;
          }

          visited[x][y] = true;

          if (grid[x][y] == 0) {
            continue;
          }

          grid[x][y] = id;
          area++;

          for (int[] direction : directions) {
            int dx = direction[0];
            int dy = direction[1];

            int newX = x + dx;
            int newY = y + dy;

            if (newX < 0 || newX >= n || newY < 0 || newY >= n) {
              continue;
            }

            if (grid[newX][newY] == 0) {
              visited[newX][newY] = true;
              continue;
            }

            queue.offer(new int[] { newX, newY });
          }
        }

        islands.put(id, area);
      }
    }

    if (islands.isEmpty()) {
      return 1;
    }

    int max = Collections.max(islands.values());

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        if (grid[i][j] > 0) {
          continue;
        }

        int sum = 1;

        Set<Integer> neighbours = new HashSet<>();

        for (int[] direction : directions) {
          int dx = direction[0];
          int dy = direction[1];

          int x = i + dx;
          int y = j + dy;

          if (x < 0 || x >= n || y < 0 || y >= n || grid[x][y] == 0) {
            continue;
          }

          neighbours.add(grid[x][y]);
        }

        for (int id : neighbours) {
          sum += islands.getOrDefault(id, 0);
        }

        max = Math.max(max, sum);
      }
    }

    return max;
  }
}
